File: android/app/src/main/java/com/listen/app/network/ApiClient.kt
package com.listen.app.network

/**
 * listen — thin networking layer over the real backend (the backend/app/api
 * endpoints, a standalone service — see web/next.config.mjs's rewrites()
 * for how the web app reaches the same backend)
 *
 * Sessions are cookie-based (see backend/lib/auth.js). OkHttp has no built-in
 * disk-persisted cookie jar (unlike iOS's HTTPCookieStorage, which persists
 * automatically) — PersistentCookieJar below backs it with SharedPreferences
 * so a session survives app restarts.
 */

import android.content.Context
import android.content.SharedPreferences
import com.listen.app.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.SerializationException
import kotlinx.serialization.serializer
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

object ApiConfig {
    /** Set per build type in app/build.gradle.kts — debug points at the
     * local backend/ dev server (10.0.2.2:4000), release at the deployed
     * Render URL. */
    val baseUrl: String = BuildConfig.API_BASE_URL
}

class ApiException(val code: String?, message: String) : Exception(message)

@kotlinx.serialization.Serializable
@PublishedApi
internal data class ErrorBody(val error: String? = null, val message: String? = null)

/** A decoded `{ ok: true }` response — several endpoints return only this. */
@kotlinx.serialization.Serializable
data class OkResponse(val ok: Boolean)

private class PersistentCookieJar(context: Context) : CookieJar {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("listen_cookies", Context.MODE_PRIVATE)
    private val key = "session_cookie"

    @Synchronized
    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        val session = cookies.firstOrNull { it.name == "listen_session" }
        if (session != null) {
            prefs.edit().putString(key, session.toString()).apply()
        }
    }

    @Synchronized
    override fun loadForRequest(url: HttpUrl): List<Cookie> {
        val raw = prefs.getString(key, null) ?: return emptyList()
        val cookie = Cookie.parse(url, raw) ?: return emptyList()
        return listOf(cookie)
    }

    @Synchronized
    fun clear() {
        prefs.edit().remove(key).apply()
    }
}

class ApiClient(context: Context) {
    private val cookieJar = PersistentCookieJar(context.applicationContext)

    // inline reified send()/upload() below need to reach these across
    // module boundaries, so they can't stay plain `private`.
    @PublishedApi
    internal val client = OkHttpClient.Builder()
        .cookieJar(cookieJar)
        // TTS generation calls a real provider and can legitimately take a
        // while — the OkHttp default (10s) is too tight for that, even
        // though it's fine for everything else.
        .callTimeout(120, TimeUnit.SECONDS)
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .build()

    @PublishedApi
    internal val json = Json { ignoreUnknownKeys = true }

    fun clearSession() = cookieJar.clear()

    suspend inline fun <reified T> get(path: String): T =
        send(path, "GET", null)

    suspend inline fun <reified T> post(path: String): T =
        send(path, "POST", null)

    suspend inline fun <reified B, reified T> post(path: String, body: B): T =
        send(path, "POST", json.encodeToString(serializer<B>(), body))

    suspend inline fun <reified B, reified T> patch(path: String, body: B): T =
        send(path, "PATCH", json.encodeToString(serializer<B>(), body))

    suspend inline fun <reified B, reified T> put(path: String, body: B): T =
        send(path, "PUT", json.encodeToString(serializer<B>(), body))

    suspend inline fun <reified B, reified T> delete(path: String, body: B): T =
        send(path, "DELETE", json.encodeToString(serializer<B>(), body))

    suspend inline fun <reified T> delete(path: String): T =
        send(path, "DELETE", null)

    suspend inline fun <reified T> send(path: String, method: String, jsonBody: String?): T =
        withContext(Dispatchers.IO) {
            val requestBuilder = Request.Builder().url(ApiConfig.baseUrl + path)
            val body = jsonBody?.toRequestBody("application/json".toMediaType())
            when (method) {
                "GET" -> requestBuilder.get()
                "DELETE" -> if (body != null) requestBuilder.delete(body) else requestBuilder.delete()
                else -> requestBuilder.method(method, body ?: "".toRequestBody("application/json".toMediaType()))
            }

            val response = try {
                client.newCall(requestBuilder.build()).execute()
            } catch (e: Exception) {
                throw ApiException(null, "Check your connection and try again.")
            }

            response.use {
                val text = it.body?.string().orEmpty()
                if (!it.isSuccessful) {
                    val err = runCatching { json.decodeFromString<ErrorBody>(text) }.getOrNull()
                    throw ApiException(err?.error, err?.message ?: "Something went wrong (${it.code}).")
                }
                try {
                    json.decodeFromString(serializer<T>(), text)
                } catch (e: SerializationException) {
                    throw ApiException(null, "Got an unexpected response. Try again.")
                }
            }
        }

    /** Multipart upload — used by POST /api/upload. */
    suspend inline fun <reified T> upload(path: String, fileBytes: ByteArray, filename: String, mimeType: String): T =
        withContext(Dispatchers.IO) {
            val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart(
                    "file", filename,
                    fileBytes.toRequestBody(mimeType.toMediaType())
                )
                .build()

            val request = Request.Builder().url(ApiConfig.baseUrl + path).post(body).build()

            val response = try {
                client.newCall(request).execute()
            } catch (e: Exception) {
                throw ApiException(null, "Check your connection and try again.")
            }

            response.use {
                val text = it.body?.string().orEmpty()
                if (!it.isSuccessful) {
                    val err = runCatching { json.decodeFromString<ErrorBody>(text) }.getOrNull()
                    throw ApiException(err?.error, err?.message ?: "Could not read that file.")
                }
                try {
                    json.decodeFromString(serializer<T>(), text)
                } catch (e: SerializationException) {
                    throw ApiException(null, "Got an unexpected response. Try again.")
                }
            }
        }

    /** Streams raw bytes — used to download generated audio for local playback checks. */
    suspend fun getBytes(path: String): ByteArray = withContext(Dispatchers.IO) {
        val request = Request.Builder().url(ApiConfig.baseUrl + path).get().build()
        client.newCall(request).execute().use {
            if (!it.isSuccessful) throw ApiException(null, "Could not download that file.")
            it.body?.bytes() ?: ByteArray(0)
        }
    }
}

File: ios/Listen/Services/APIClient.swift
//
//  APIClient.swift
//  listen — thin networking layer over the real backend (backend/app/api/**,
//  a standalone service — see backend/README or web/next.config.mjs's
//  rewrites() for how the web app reaches the same backend)
//
//  Sessions are cookie-based (see backend/lib/auth.js) — URLSession's default
//  configuration handles Set-Cookie/Cookie automatically via
//  HTTPCookieStorage, the same way a browser does, so no token plumbing
//  is needed here.
//

import Foundation

enum APIConfig {
    #if DEBUG
    /// The iOS Simulator shares the host Mac's network namespace, so
    /// "localhost" reaches the same `npm run dev` server (backend/, port
    /// 4000) the web app proxies to. A physical device needs the Mac's LAN
    /// IP instead — swap this one line, nothing else changes.
    static let baseURL = URL(string: "http://localhost:4000")!
    #else
    /// TODO: fill in once the backend is deployed — see backend/render.yaml.
    /// This is the standalone backend's own Render URL, not the web app's.
    static let baseURL = URL(string: "https://REPLACE-ME.onrender.com")!
    #endif
}

struct APIClientError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

private struct ErrorBody: Decodable {
    let error: String?
    let message: String?
}

private struct EmptyBody: Encodable {}

final class APIClient {
    static let shared = APIClient()

    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    private init() {
        let config = URLSessionConfiguration.default
        config.httpCookieStorage = .shared
        config.httpShouldSetCookies = true
        config.httpCookieAcceptPolicy = .always
        // TTS generation (POST /api/documents/:id/audio) calls a real
        // provider and can legitimately take a while — the default 60s
        // request timeout is too tight for that, even though it's fine
        // for everything else.
        config.timeoutIntervalForRequest = 120
        session = URLSession(configuration: config)

        decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601

        encoder = JSONEncoder()
    }

    /// GET (or any body-less method).
    func request<T: Decodable>(_ path: String, method: String = "GET") async throws -> T {
        try await send(path, method: method, bodyData: nil)
    }

    /// Any method with a JSON body.
    func request<T: Decodable, B: Encodable>(_ path: String, method: String = "POST", body: B) async throws -> T {
        let data = try encoder.encode(body)
        return try await send(path, method: method, bodyData: data)
    }

    private func send<T: Decodable>(_ path: String, method: String, bodyData: Data?) async throws -> T {
        var req = URLRequest(url: APIConfig.baseURL.appendingPathComponent(path))
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = bodyData

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw APIClientError(message: "Check your connection and try again.")
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIClientError(message: "Something went wrong. Try again.")
        }

        guard (200...299).contains(http.statusCode) else {
            let body = try? decoder.decode(ErrorBody.self, from: data)
            throw APIClientError(message: body?.message ?? "Something went wrong (\(http.statusCode)).")
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIClientError(message: "Got an unexpected response. Try again.")
        }
    }

    // MARK: - Multipart upload (POST /api/upload)

    func upload<T: Decodable>(_ path: String, fileData: Data, filename: String, mimeType: String) async throws -> T {
        let boundary = "Boundary-\(UUID().uuidString)"
        var req = URLRequest(url: APIConfig.baseURL.appendingPathComponent(path))
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append(
            "Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n"
                .data(using: .utf8)!
        )
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(fileData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.upload(for: req, from: body)
        } catch {
            throw APIClientError(message: "Check your connection and try again.")
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIClientError(message: "Something went wrong. Try again.")
        }
        guard (200...299).contains(http.statusCode) else {
            let errBody = try? decoder.decode(ErrorBody.self, from: data)
            throw APIClientError(message: errBody?.message ?? "Could not read that file.")
        }
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIClientError(message: "Got an unexpected response. Try again.")
        }
    }
}

/// A decoded `{ ok: true }` response — several endpoints return only this.
struct OKResponse: Decodable { let ok: Bool }

extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
